zebrad/src/components/sync.rs

```diff
@@ -550,76 +557,103 @@ where
             state_tip = ?self.latest_chain_tip.best_tip_height(),
             "starting sync, obtaining new tips"
         );
-        let mut extra_hashes = self.obtain_tips().await.map_err(|e| {
-            info!("temporary error obtaining tips: {:#}", e);
-            e
-        })?;
+        let mut extra_hashes = timeout(SYNC_RESTART_DELAY, self.obtain_tips())
+            .await
+            .map_err(Into::into)
+            // TODO: replace with flatten() when it stabilises (#70142)
+            .and_then(convert::identity)
+            .map_err(|e| {
+                info!("temporary error obtaining tips: {:#}", e);
+                e
+            })?;
         self.update_metrics();
 
         while !self.prospective_tips.is_empty() || !extra_hashes.is_empty() {
-            // Check whether any block tasks are currently ready:
-            while let Poll::Ready(Some(rsp)) = futures::poll!(self.downloads.next()) {
-                self.handle_block_response(rsp)?;
-            }
-            self.update_metrics();
+            // Avoid hangs due to service readiness or other internal operations
+            extra_hashes = timeout(BLOCK_VERIFY_TIMEOUT, self.try_to_sync_once(extra_hashes))
+                .await
+                .map_err(Into::into)
+                // TODO: replace with flatten() when it stabilises (#70142)
+                .and_then(convert::identity)?;
+        }
```

Comment:
This is the only change in the diff for this loop, the rest is code movement and removing an indent.